src/main/java/ch/njol/skript/expressions/arithmetic/ExprArithmetic.java

```diff
@@ -48,147 +49,190 @@
+		Skript.registerExpression(ExprArithmetic.class, Object.class, ExpressionType.PATTERN_MATCHES_EVERYTHING, patterns.getPatterns());
 	}
-	
-	@SuppressWarnings("null")
-	private Expression<? extends Number> first;
+
 	@SuppressWarnings("null")
-	private Expression<? extends Number> second;
+	private Expression<?> first, second;
 	@SuppressWarnings("null")
 	private Operator op;
-	
+
 	@SuppressWarnings("null")
-	private Class<? extends Number> returnType;
-	
+	private Class<?> returnType;
+
 	// A chain of expressions and operators, alternating between the two. Always starts and ends with an expression.
 	private final List<Object> chain = new ArrayList<>();
-	
+
 	// A parsed chain, like a tree
-	private ArithmeticGettable arithmeticGettable;
-	
-	@SuppressWarnings({"unchecked", "null"})
+	@Nullable
+	private ArithmeticGettable<?> arithmeticGettable;
+
+	private boolean leftGrouped, rightGrouped;
+
 	@Override
+	@SuppressWarnings({"null"})
 	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		first = (Expression<? extends Number>) exprs[0];
-		second = (Expression<? extends Number>) exprs[1];
-		
+		first = LiteralUtils.defendExpression(exprs[0]);
+		second = LiteralUtils.defendExpression(exprs[1]);
+
+		if (!LiteralUtils.canInitSafely(first, second))
+			return false;
+
+		returnType = first.getReturnType();
+		Class<?> secondClass = second.getReturnType();
+
 		PatternInfo patternInfo = patterns.getInfo(matchedPattern);
 		op = patternInfo.operator;
-		
-		if (op == Operator.DIV || op == Operator.EXP) {
-			returnType = Double.class;
-		} else {
-			Class<?> firstReturnType = first.getReturnType();
-			Class<?> secondReturnType = second.getReturnType();
-			
-			boolean firstIsInt = false;
-			boolean secondIsInt = false;
-			for (final Class<?> i : INTEGER_CLASSES) {
-				firstIsInt |= i.isAssignableFrom(firstReturnType);
-				secondIsInt |= i.isAssignableFrom(secondReturnType);
+		leftGrouped = patternInfo.leftGrouped;
+		rightGrouped = patternInfo.rightGrouped;
+
+		if (Number.class.isAssignableFrom(returnType)) {
+			if (op == Operator.DIV || op == Operator.EXP) {
+				returnType = Double.class;
+			} else {
+				Class<?> firstReturnType = first.getReturnType();
+				Class<?> secondReturnType = second.getReturnType();
+
+				boolean firstIsInt = false;
+				boolean secondIsInt = false;
+				for (final Class<?> i : INTEGER_CLASSES) {
+					firstIsInt |= i.isAssignableFrom(firstReturnType);
+					secondIsInt |= i.isAssignableFrom(secondReturnType);
+				}
+
+				returnType = firstIsInt && secondIsInt ? Long.class : Double.class;
 			}
-			
-			returnType = firstIsInt && secondIsInt ? Long.class : Double.class;
 		}
-		
+
+		Arithmetic<?> arithmetic = null;
+
+		if (!returnType.equals(Object.class)
+				&& !secondClass.equals(Object.class)
+				&& (arithmetic = getArithmetic(returnType, op, secondClass)) == null) {
+			return false;
+		}
```

Comment:
This is if two types aren't compatible for arithmetic right? If so, I think we should have a custom error message.

Reply:
What should the error message be?